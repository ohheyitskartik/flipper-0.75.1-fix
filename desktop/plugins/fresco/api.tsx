/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

export type ImageId = string;

// Listing images

export type CacheInfo = {
  cacheType: string;
  clearKey?: string; // set this if this cache level supports clear(<key>)
  sizeBytes: number;
  maxSizeBytes?: number;
  imageIds: Array<ImageId>;
};

export type ImagesList = Array<CacheInfo>;

// The iOS Flipper api does not support a top-level array, so we wrap it in an object
export type ImagesListResponse = {
  levels: ImagesList;
};

// listImages() -> ImagesListResponse

// Getting details on a specific image

export type ImageBytes = string;

export type ImageData = {
  imageId: ImageId;
  uri?: string;
  width: number;
  height: number;
  sizeBytes: number;
  data: ImageBytes;
  surface?: string;
};

// getImage({imageId: string}) -> ImageData

// Subscribing to image events (requests and prefetches)

export type Timestamp = number;

export type ViewportData = {
  width: number;
  height: number;
  scanDisplayTime: {[scan_number: number]: Timestamp};
};

export type ImageEvent = {
  imageIds: Array<ImageId>;
  attribution: Array<string>;
  startTime: Timestamp;
  endTime: Timestamp;
  source: string;
  coldStart: boolean;
  viewport?: ViewportData; // not set for prefetches
};

// Misc

export type FrescoDebugOverlayEvent = {
  enabled: boolean;
};

export type AndroidCloseableReferenceLeakEvent = {
  identityHashCode: string;
  className: string;
  stacktrace: string | null;
};
