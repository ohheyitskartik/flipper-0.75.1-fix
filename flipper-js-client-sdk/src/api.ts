/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

export type FlipperPluginID = string;

export type FlipperMethodID = string;

export class FlipperResponder {
  messageID?: number;
  private client: FlipperClient;

  constructor(messageID: number, client: FlipperClient) {
    this.messageID = messageID;
    this.client = client;
  }

  success(response?: any) {
    this.client.sendData({id: this.messageID, success: response});
  }

  error(response?: any) {
    this.client.sendData({id: this.messageID, error: response});
  }
}

export type FlipperReceiver = (
  params: any,
  responder: FlipperResponder,
) => void;

export class FlipperConnection {
  pluginId: FlipperPluginID;
  private client: FlipperClient;
  private subscriptions: Map<FlipperMethodID, FlipperReceiver> = new Map();

  constructor(pluginId: FlipperPluginID, client: FlipperClient) {
    this.pluginId = pluginId;
    this.client = client;
  }

  send(method: FlipperMethodID, params: any) {
    this.client.sendData({
      method: 'execute',
      params: {
        api: this.pluginId,
        method,
        params,
      },
    });
  }

  receive(method: FlipperMethodID, receiver: FlipperReceiver) {
    this.subscriptions.set(method, receiver);
  }

  call(method: FlipperMethodID, params: any, responder: FlipperResponder) {
    const receiver = this.subscriptions.get(method);
    if (receiver == null) {
      const errorMessage = `Receiver ${method} not found.`;
      responder.error({message: errorMessage});
      return;
    }
    receiver.call(receiver, params, responder);
  }

  hasReceiver(method: FlipperMethodID): boolean {
    return this.subscriptions.has(method);
  }
}

export interface FlipperPlugin {
  /**
   * @return The id of this plugin. This is the namespace which Flipper desktop plugins will call
   *     methods on to route them to your plugin. This should match the id specified in your React
   *     plugin.
   */
  getId(): string;

  /**
   * Called when a connection has been established. The connection passed to this method is valid
   * until {@link FlipperPlugin#onDisconnect()} is called.
   */
  onConnect(connection: FlipperConnection): void;

  /**
   * Called when the connection passed to `FlipperPlugin#onConnect(FlipperConnection)` is no
   * longer valid. Do not try to use the connection in or after this method has been called.
   */
  onDisconnect(): void;

  /**
   * Returns true if the plugin is meant to be run in background too, otherwise it returns false.
   */
  runInBackground(): boolean;
}

export abstract class FlipperClient {
  private _isConnected: boolean = false;
  protected plugins: Map<FlipperPluginID, FlipperPlugin> = new Map();
  protected connections: Map<FlipperPluginID, FlipperConnection> = new Map();

  addPlugin(plugin: FlipperPlugin) {
    if (this._isConnected) {
      this.connectPlugin(plugin);
    }
    this.plugins.set(plugin.getId(), plugin);
  }

  getPlugin(id: FlipperPluginID): FlipperPlugin | undefined {
    return this.plugins.get(id);
  }

  onConnect() {
    if (this._isConnected) {
      return;
    }
    this._isConnected = true;
  }

  onDisconnect() {
    this._isConnected = false;
    for (const plugin of this.plugins.values()) {
      this.disconnectPlugin(plugin);
    }
  }

  abstract start(appName: string): void;

  abstract stop(): void;

  abstract sendData(payload: any): void;

  abstract isAvailable(): boolean;

  protected onMessageReceived(message: {
    method: string;
    id: number;
    params: any;
  }) {
    let responder: FlipperResponder | undefined;
    try {
      const {method, params, id} = message;
      responder = new FlipperResponder(id, this);

      if (method === 'getPlugins') {
        responder.success({plugins: [...this.plugins.keys()]});
        return;
      }

      if (method === 'getBackgroundPlugins') {
        responder.success({
          plugins: [...this.plugins.keys()].filter((key) =>
            this.plugins.get(key)?.runInBackground(),
          ),
        });
        return;
      }

      if (method === 'init') {
        const identifier = params['plugin'] as string;
        const plugin = this.plugins.get(identifier);
        if (plugin == null) {
          const errorMessage = `Plugin ${identifier} not found for method ${method}`;
          responder.error({message: errorMessage, name: 'PluginNotFound'});
          return;
        }

        this.connectPlugin(plugin);
        return;
      }

      if (method === 'deinit') {
        const identifier = params['plugin'] as string;
        const plugin = this.plugins.get(identifier);
        if (plugin == null) {
          const errorMessage = `Plugin ${identifier} not found for method ${method}`;
          responder.error({message: errorMessage, name: 'PluginNotFound'});
          return;
        }

        this.disconnectPlugin(plugin);
        return;
      }

      if (method === 'execute') {
        const identifier = params['api'] as string;
        const connection = this.connections.get(identifier);
        if (connection == null) {
          const errorMessage = `Connection ${identifier} not found for plugin identifier`;

          responder.error({message: errorMessage, name: 'ConnectionNotFound'});
          return;
        }

        connection.call(
          params['method'] as string,
          params['params'],
          responder,
        );
        return;
      }

      if (method === 'isMethodSupported') {
        const identifier = params['api'].getString();
        const connection = this.connections.get(identifier);
        if (connection == null) {
          const errorMessage = `Connection ${identifier} not found for plugin identifier`;

          responder.error({message: errorMessage, name: 'ConnectionNotFound'});
          return;
        }
        const isSupported = connection.hasReceiver(
          params['method'].getString(),
        );
        responder.success({isSupported: isSupported});
        return;
      }

      const response = {message: 'Received unknown method: ' + method};
      responder.error(response);
    } catch (e) {
      if (responder) {
        responder.error({
          message: 'Unknown error during ' + JSON.stringify(message),
          name: 'Unknown',
        });
      }
    }
  }

  private connectPlugin(plugin: FlipperPlugin): void {
    const id = plugin.getId();
    const connection = new FlipperConnection(id, this);
    plugin.onConnect(connection);
    this.connections.set(id, connection);
  }

  private disconnectPlugin(plugin: FlipperPlugin): void {
    const id = plugin.getId();
    plugin.onDisconnect();
    this.connections.delete(id);
  }
}
