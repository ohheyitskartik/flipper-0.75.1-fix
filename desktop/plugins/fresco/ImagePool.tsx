/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {ImageId, ImageData} from './api';

export type ImagesMap = {[imageId in ImageId]: ImageData};

const maxInflightRequests = 10;

export default class ImagePool {
  cache: ImagesMap = {};
  requested: {[imageId in ImageId]: boolean} = {};
  queued: Array<ImageId> = [];
  inFlightRequests: number = 0;
  fetchImage: (imageId: ImageId) => void;
  updateNotificationScheduled: boolean = false;
  onPoolUpdated: (images: ImagesMap) => void;

  constructor(
    fetchImage: (imageId: ImageId) => void,
    onPoolUpdated: (images: ImagesMap) => void,
  ) {
    this.fetchImage = fetchImage;
    this.onPoolUpdated = onPoolUpdated;
  }

  getImages(): ImagesMap {
    return {...this.cache};
  }

  fetchImages(ids: Array<string>) {
    for (const id of ids) {
      if (!this.cache[id] && !this.requested[id]) {
        this.requested[id] = true;

        if (this.inFlightRequests < maxInflightRequests) {
          this.inFlightRequests++;
          this.fetchImage(id);
        } else {
          this.queued.unshift(id);
        }
      }
    }
  }

  clear() {
    this.cache = {};
    this.requested = {};
  }

  _fetchCompleted(image: ImageData): void {
    this.cache[image.imageId] = image;
    delete this.requested[image.imageId];

    if (this.queued.length > 0) {
      const popped = this.queued.pop() as string;
      this.fetchImage(popped);
    } else {
      this.inFlightRequests--;
    }

    if (!this.updateNotificationScheduled) {
      this.updateNotificationScheduled = true;
      window.setTimeout(this._notify, 1000);
    }
  }

  _notify = () => {
    this.updateNotificationScheduled = false;
    this.onPoolUpdated(this.getImages());
  };
}
