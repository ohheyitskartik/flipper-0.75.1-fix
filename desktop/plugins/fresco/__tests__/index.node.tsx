/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import FrescoPlugin from '../index';
import {PersistedState, ImageEventWithId} from '../index';
import {AndroidCloseableReferenceLeakEvent} from '../api';
import {Notification} from 'flipper';
import {ImagesMap} from '../ImagePool';

type ScanDisplayTime = {[scan_number: number]: number};

function mockPersistedState(
  imageSizes: Array<{
    width: number;
    height: number;
  }> = [],
  viewport: {
    width: number;
    height: number;
  } = {width: 150, height: 150},
): PersistedState {
  const scanDisplayTime: ScanDisplayTime = {};
  scanDisplayTime[1] = 3;
  const events: Array<ImageEventWithId> = [
    {
      imageIds: [...Array(imageSizes.length).keys()].map(String),
      eventId: 0,
      attribution: [],
      startTime: 1,
      endTime: 2,
      source: 'source',
      coldStart: true,
      viewport: {...viewport, scanDisplayTime},
    },
  ];

  const imagesMap = imageSizes.reduce((acc, val, index) => {
    acc[index] = {
      imageId: String(index),
      width: val.width,
      height: val.height,
      sizeBytes: 10,
      data: 'undefined',
    };
    return acc;
  }, {} as ImagesMap);

  return {
    surfaceList: new Set(),
    images: [],
    events,
    imagesMap,
    closeableReferenceLeaks: [],
    isLeakTrackingEnabled: false,
    nextEventId: 0,
    showDiskImages: false,
  };
}

test('notifications for leaks', () => {
  const notificationReducer: (
    persistedState: PersistedState,
  ) => Array<Notification> = FrescoPlugin.getActiveNotifications;
  const closeableReferenceLeaks: Array<AndroidCloseableReferenceLeakEvent> = [
    {
      identityHashCode: 'deadbeef',
      className: 'com.facebook.imagepipeline.memory.NativeMemoryChunk',
      stacktrace: null,
    },
    {
      identityHashCode: 'f4c3b00c',
      className: 'com.facebook.flipper.SomeMemoryAbstraction',
      stacktrace: null,
    },
  ];
  const persistedStateWithoutTracking = {
    ...mockPersistedState(),
    closeableReferenceLeaks,
    isLeakTrackingEnabled: false,
  };
  const emptyNotifs = notificationReducer(persistedStateWithoutTracking);
  expect(emptyNotifs).toHaveLength(0);

  const persistedStateWithTracking = {
    ...mockPersistedState(),
    closeableReferenceLeaks,
    isLeakTrackingEnabled: true,
  };
  const notifs = notificationReducer(persistedStateWithTracking);
  expect(notifs).toHaveLength(2);
  expect(notifs[0].message).toMatchSnapshot();
  expect(notifs[1].title).toContain('SomeMemoryAbstraction');
});
