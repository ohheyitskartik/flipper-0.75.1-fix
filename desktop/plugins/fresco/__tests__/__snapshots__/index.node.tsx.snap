// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notifications for leaks 1`] = `
<React.Fragment>
  <Styled(div)>
    CloseableReference leaked for
     
    <Text
      code={true}
    >
      com.facebook.imagepipeline.memory.NativeMemoryChunk
    </Text>
    (identity hashcode: 
    deadbeef
    ).
  </Styled(div)>
  <Styled(div)>
    <Text
      bold={true}
    >
      Stacktrace:
    </Text>
  </Styled(div)>
  <Styled(div)>
    <Text
      code={true}
    >
      &lt;unavailable&gt;
    </Text>
  </Styled(div)>
</React.Fragment>
`;
